File: common/src/main/java/dev/mayaqq/estrogen/registry/advancements/triggers/InsertJarTrigger.java
package dev.mayaqq.estrogen.registry.advancements.triggers;

import com.google.gson.JsonObject;
import dev.mayaqq.estrogen.Estrogen;
import net.minecraft.advancements.critereon.*;
import net.minecraft.resources.ResourceLocation;
import net.minecraft.server.level.ServerPlayer;

public class InsertJarTrigger extends SimpleCriterionTrigger<InsertJarTrigger.TriggerInstance> {
    static final ResourceLocation ID = Estrogen.id("insert_jar");

    @Override
    protected TriggerInstance createInstance(JsonObject json, ContextAwarePredicate predicate, DeserializationContext deserializationContext) {
        return new TriggerInstance(predicate);
    }

    public ResourceLocation getId() {
        return ID;
    }

    public void trigger(ServerPlayer player) {
        this.trigger(player, (instance) -> true);
    }

    public static class TriggerInstance extends AbstractCriterionTriggerInstance {
        public TriggerInstance(ContextAwarePredicate player) {
            super(InsertJarTrigger.ID, player);
        }

        public static TriggerInstance insertJar() {
            return new TriggerInstance(ContextAwarePredicate.ANY);
        }

        public JsonObject serializeToJson(SerializationContext context) {
            return super.serializeToJson(context);
        }
    }
}

File: src/main/kotlin/dev/mayaqq/estrogen/content/advancements/triggers/InsertJarTrigger.kt
package dev.mayaqq.estrogen.content.advancements.triggers

import com.mojang.serialization.Codec
import com.mojang.serialization.codecs.RecordCodecBuilder
import dev.mayaqq.estrogen.content.advancements.triggers.InsertJarTrigger.TriggerInstance
import net.minecraft.advancements.critereon.*
import net.minecraft.server.level.ServerPlayer
import java.util.Optional

class InsertJarTrigger : SimpleCriterionTrigger<TriggerInstance>() {
    override fun codec(): Codec<TriggerInstance> = TriggerInstance.CODEC

    fun trigger(player: ServerPlayer) {
        this.trigger(player) { instance: TriggerInstance ->
            true
        }
    }

    @JvmRecord
    data class TriggerInstance(val player: ContextAwarePredicate) : SimpleInstance {
        override fun player(): Optional<ContextAwarePredicate> = Optional.of(player)

        companion object {
            val CODEC: Codec<TriggerInstance> = RecordCodecBuilder.create { instance -> instance.group(
                    EntityPredicate.ADVANCEMENT_CODEC.fieldOf("player").forGetter(TriggerInstance::player)
                ).apply(instance, ::TriggerInstance)
            }

            fun insertJar() : TriggerInstance {
                return TriggerInstance(ContextAwarePredicate.create())
            }
        }
    }
}